File: code/packages/python/macsyma-parser/tests/test_parser.py
"""MACSYMA parser tests.

These verify that the grammar-driven parser produces the expected
``ASTNode`` shapes for representative MACSYMA programs. We don't pin
down every child position (that would make the tests brittle against
grammar tweaks); instead we check structural properties — root rule
name, number of statements, key nonterminal types appearing.
"""

from __future__ import annotations

from lang_parser import ASTNode, find_nodes
from macsyma_parser import parse_macsyma


def test_single_expression_statement() -> None:
    ast = parse_macsyma("x;")
    assert ast.rule_name == "program"
    statements = find_nodes(ast, "statement")
    assert len(statements) == 1


def test_arithmetic_expression() -> None:
    ast = parse_macsyma("1 + 2 * 3;")
    # Expect at least one `additive` and one `multiplicative` node,
    # since the grammar's precedence cascade uses both.
    assert len(find_nodes(ast, "additive")) >= 1
    assert len(find_nodes(ast, "multiplicative")) >= 1


def test_power_expression() -> None:
    ast = parse_macsyma("x^2;")
    powers = find_nodes(ast, "power")
    assert len(powers) >= 1


def test_parenthesized_expression() -> None:
    # `(1 + 2) * 3` should force the `additive` inside `multiplicative`
    # via the explicit `group` rule.
    ast = parse_macsyma("(1 + 2) * 3;")
    assert len(find_nodes(ast, "group")) >= 1


def test_function_call() -> None:
    ast = parse_macsyma("f(x, y);")
    postfix_nodes = find_nodes(ast, "postfix")
    assert len(postfix_nodes) >= 1


def test_assignment() -> None:
    ast = parse_macsyma("a : 5;")
    assign_nodes = find_nodes(ast, "assign")
    assert len(assign_nodes) >= 1


def test_function_definition() -> None:
    ast = parse_macsyma("f(x) := x^2;")
    assert ast.rule_name == "program"
    # The `assign` rule handles both `:` and `:=`, so we expect an
    # assign node in the tree.
    assert len(find_nodes(ast, "assign")) >= 1


def test_list_literal() -> None:
    ast = parse_macsyma("[1, 2, 3];")
    lists = find_nodes(ast, "list")
    assert len(lists) >= 1


def test_multiple_statements() -> None:
    ast = parse_macsyma("a : 5; b : 10; a + b;")
    statements = find_nodes(ast, "statement")
    assert len(statements) == 3


def test_suppress_output_dollar() -> None:
    # Both `;` and `$` are valid terminators.
    ast = parse_macsyma("x$")
    assert ast.rule_name == "program"
    assert len(find_nodes(ast, "statement")) == 1


def test_mixed_terminators() -> None:
    ast = parse_macsyma("a : 5$ b : 10; a + b;")
    assert len(find_nodes(ast, "statement")) == 3


def test_nested_function_calls() -> None:
    ast = parse_macsyma("diff(f(x), x);")
    # f(x) is one postfix with a call suffix; diff(...) is another.
    postfix_nodes = find_nodes(ast, "postfix")
    assert len(postfix_nodes) >= 2


def test_percent_constants() -> None:
    ast = parse_macsyma("%pi + %e;")
    # No crash; the atoms should parse as NAME tokens.
    assert ast.rule_name == "program"


def test_comparison() -> None:
    ast = parse_macsyma("x = 4;")
    assert len(find_nodes(ast, "comparison")) >= 1


def test_logical_operators() -> None:
    ast = parse_macsyma("a and b or not c;")
    assert len(find_nodes(ast, "logical_or")) >= 1
    assert len(find_nodes(ast, "logical_and")) >= 1


def test_unary_minus() -> None:
    ast = parse_macsyma("-x;")
    assert len(find_nodes(ast, "unary")) >= 1


def test_comment_ignored() -> None:
    ast = parse_macsyma("/* a note */ 42;")
    assert len(find_nodes(ast, "statement")) == 1


def test_empty_program() -> None:
    # Zero statements is valid — the grammar uses `{ statement }`.
    ast = parse_macsyma("")
    assert ast.rule_name == "program"
    assert len(find_nodes(ast, "statement")) == 0


def test_returns_ast_node() -> None:
    ast = parse_macsyma("1;")
    assert isinstance(ast, ASTNode)

File: code/packages/rust/macsyma-parser/tests/test_parser.rs
use coding_adventures_macsyma_parser::parse_macsyma;
use parser::grammar_parser::{ASTNodeOrToken, GrammarASTNode};

fn count_rule(node: &GrammarASTNode, rule_name: &str) -> usize {
    let mut count = usize::from(node.rule_name == rule_name);
    for child in &node.children {
        if let ASTNodeOrToken::Node(child_node) = child {
            count += count_rule(child_node, rule_name);
        }
    }
    count
}

#[test]
fn parses_single_expression_statement() {
    let ast = parse_macsyma("x;");
    assert_eq!(ast.rule_name, "program");
    assert_eq!(count_rule(&ast, "statement"), 1);
}

#[test]
fn parses_precedence_bearing_arithmetic() {
    let ast = parse_macsyma("1 + 2 * 3;");
    assert!(count_rule(&ast, "additive") > 0);
    assert!(count_rule(&ast, "multiplicative") > 0);
}

#[test]
fn parses_function_definitions_and_calls() {
    let ast = parse_macsyma("f(x) := x^2; diff(f(x), x);");
    assert_eq!(count_rule(&ast, "statement"), 2);
    assert!(count_rule(&ast, "postfix") >= 2);
}

#[test]
fn parses_lists_comparisons_logic_and_dollar_terminators() {
    let ast = parse_macsyma("[1, 2, 3]$ a < b and not false;");
    assert!(count_rule(&ast, "list") > 0);
    assert!(count_rule(&ast, "comparison") > 0);
    assert!(count_rule(&ast, "logical_and") > 0);
}

#[test]
fn parses_control_flow_grammar_forms() {
    let ast = parse_macsyma("if x < 0 then -x else x; while x < 3 do x : x + 1;");
    assert_eq!(count_rule(&ast, "if_expr"), 1);
    assert_eq!(count_rule(&ast, "while_expr"), 1);
}
